### DISCUSSION TITLE
SSH connection to IPSEC gateway breaks once VPN connection established.

### DISCUSSION THREAD

**Discussion by hombibi** [strongswan/strongswan#1103] [Q&A]
I am trying to setup a headless ubuntu server as ipsec vpn gateway for my network, with a connection to a vps service provider.

Since I restarted more then once I created a rudimentary script to setup the ubuntu server (default setup, two nics, no specific packes installed to start with), as router, vpn client and gateway for my local network. See below. I have a few issues which I am not able to solve myself.
1) My SSH connection to the server (or vpn client if you may) is interrupted as soon as the connection establishes, and this prevents me from completing the setup.
2) In the connection log I find "handling INTERNAL_IP4_NETMASK attribute failed" after setting the dns server in /etc/resolv.conf which I am not sure how to solv.
3) The connection seems to work but forwarding the local network fails, although due to nr 1) I have some difficulty to find out why.

Below the server/router/vpn client setup: 

```
  1 #!/bin/bash
  2
  3 ###############################################################################
  4 #
  5 #
  6 # Setup of ubuntu router after initial installation with two network interfaces.
  7 #
  8 #
  9 ###############################################################################
 10
 11 iflan=ens160 #192.6.7.8
 12 ifwan=ens192 #192.2.3.4
 13 ipt="$(which iptables)" # only use which in bash, command -v is POSIX compliant
 14 printf "iptables = ${ipt}"
 15
 16 exists()
 17 {
 18   command -v "$1" >/dev/null 2>&1
 19 }
 20
 21 printf "\n\n\n"
 22 printf "#####################################################################\n"
 23 printf "\nThis script tries to setup the current machine as gateway via an\n"
 24 printf "IPSec vpn connection from a basis ubuntu 20.04 installation.i\n\n"
 25 printf "#####################################################################\n"
 26
 27
 28 # remove Cloud Init, unnecessary for local installation
 29 if exists cloud-init;
 30 then
 31    printf "\n* Starting to remove cloud init\n"
 32    sudo apt-get autoremove --purge cloud-init -y
 33    sudo rm -rf /etc/cloud/ && sudo rm -rf /var/lib/cloud/
 34 else
 35    printf "\n* Cloud-init already removed, no action.\n"
 36 fi
 37
 38 if exists ifconfig;
 39 then
 40    printf "\n* Net-tools already installed, no action. \n"
 41 else
 42    printf "\n* No ifconfig found, installing net-tools\n"
 43    sudo apt-get install net-tools -y
 44 fi
 45
 46 printf "\n* Who is listening:\n\n"
 47 sudo netstat -tulpn | grep LISTEN
 48
 49
 50 if [[ $(sysctl -n net.ipv4.ip_forward) == 0 ]];
 51 then
 52    printf "\n* Setting up routing\n"
 53    sudo sed -i 's/#net.ipv4.ip_forward=1/net.ipv4.ip_forward=1/' /etc/sysctl.conf
 54    sudo sysctl -p
 55 else
 56    printf "\n* Forwarding already set, no change made.\n"
 57 fi
 58
 59 #disable ipv6 to avoid leaking though ipv6
 60 #Check if already disabled:
 61 sudo grep -q 'Disable IPV6 to avoid leaking besides vpn tunnel' /etc/sysctl.conf
 62 if [[ $? -eq 0 ]]
 63 then
 64    printf "\n* IPv6 already disabled, no changes to be made."
 65 else
 66
 67 cat << EOF >> /etc/sysctl.conf
 68 ################################################################################
 69 # Disable IPV6 to avoid leaking besides vpn tunnel
 70 ################################################################################
 71 net.ipv6.conf.all.disable_ipv6 = 1
 72 net.ipv6.conf.default.disable_ipv6 = 1
 73 net.ipv6.conf.lo.disable_ipv6 = 1
 74 #net.ipv6.conf.tun0.disable_ipv6 = 1
 75 EOF
 76 fi
 77
 78 # reload sysctl.conf
 79 sudo sysctl -p
 80
 81 printf "\n* Firewall nat and forward."
 82 printf "\n* Current Firewall settings:"
 83 for table in $(</proc/net/ip_tables_names)
 84 do
 85    printf "\n${table}\n"
 86    sudo "${ipt}" -S -t ${table}
 87 done
 88 read -p "Do you want to continue, starting with flushing all rules? " -n 1 -r
 89 #1read -p "Do you want to continue, starting with flushing all rules? " -n 1 -r -i y -e
 90 # -n 1 accepts only one character, and -r reads literally otherwise read would see
 91 # the backslash as an escape and and wait for the second character.
 92 # $REPLY is the default variable unless another is provided. (which is not the case)
 93 # =~ is a regex match operator.
 94 # The string is defined starting with ^, either Y or y, and ending $, so nothing else.
 95 printf "\n"
 96 if [[ ! "${REPLY}" =~ ^[Yy]$ ]]
 97 then
 98    exit 1
 99 else
100    for table in $(</proc/net/ip_tables_names)
101    do
102       /usr/sbin/iptables -t ${table} -F
103       /usr/sbin/iptables -t ${table} -X
104       /usr/sbin/iptables -t ${table} -Z
105    done
106 fi
107
108
109
110 # start test firewall rules
111
112
113
114
115
116
117
118
119
120
121
122 # nothing here yet.
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137 # end test firewall rules
138
139 printf "\n* Setting routing iptables rules"
140 # packages received from lan, to be forwared through the wan interface
141 sudo "${ipt}" -A FORWARD -i "${iflan}" -o "${ifwan}" -j ACCEPT
142
143 # packages associated with existing connection, to be forwarde to lan interfaces
144 sudo "${ipt}" -A FORWARD -i "${ifwan}" -o "${iflan}" -m state --state RELATED,ESTABLISHED -j ACCEPT
145
146 # Masquerade
147 sudo "${ipt}" -t nat -I POSTROUTING -m policy --pol ipsec --dir out -j ACCEPT
148 sudo "${ipt}" -t nat -A POSTROUTING -o "${ifwan}" -j MASQUERADE
149 sudo "${ipt}" -t nat -A POSTROUTING -o "${ifwan}" -m policy --dir out --pol ipsec -j ACCEPT
150
151
152 # rules to address possible MTU/MSS packet size issues.
153 sudo "${ipt}" -t mangle -A FORWARD -m policy --pol ipsec --dir in -p tcp -m tcp --tcp-flags SYN,RST SYN -m tcp    mss --mss 1361:1536 -j TCPMSS --set-mss 1360
154 sudo "${ipt}" -t mangle -A FORWARD -m policy --pol ipsec --dir out -p tcp -m tcp --tcp-flags SYN,RST SYN -m tc    pmss --mss 1361:1536 -j TCPMSS --set-mss 1360
155
156
157 printf "\n* Current Firewall settings:"
158 for table in $(</proc/net/ip_tables_names)
159 do
160    printf "\n${table}"
161    sudo "${ipt}" -S -t ${table}
162 done
163
164 printf "\n* Installing DNSMasqd"
165 if exists dnsmasq
166 then
167    printf "\n* DNSMask already installed, no installation needed."
168 else
169    sudo systemctl disable systemd-resolved
170    sudo systemctl stop systemd-resolved
171    sudo unlink /etc/resolv.conf
172    echo nameserver 1.1.1.1 | sudo tee /etc/resolv.conf
173    sudo apt-get update
174    sudo apt-get install dnsmasq -y
175 fi
176
177 printf "\n* Writing DNSMasq.conf"
178
179 cat << EOF > /etc/dnsmasq.conf
180 # Listen on this specific port instead of the standard DNS port
181 # (53). Setting this to zero completely disables DNS function,
182 # leaving only DHCP and/or TFTP.
183 port=53
184 # Never forward plain names (without a dot or domain part)
185 domain-needed
186 # Never forward addresses in the non-routed address spaces.
187 bogus-priv
188 # By  default,  dnsmasq  will  send queries to any of the upstream
189 # servers it knows about and tries to favour servers to are  known
190 # to  be  up.  Uncommenting this forces dnsmasq to try each query
191 # with  each  server  strictly  in  the  order  they   appear   in
192 # /etc/resolv.conf
193 strict-order
194 # Set this (and domain: see below) if you want to have a domain
195 # automatically added to simple names in a hosts-file.
196 expand-hosts
197 # Set the domain for dnsmasq. this is optional, but if it is set, it
198 # does the following things.
199 # 1) Allows DHCP hosts to have fully qualified domain names, as long
200 #     as the domain part matches this setting.
201 # 2) Sets the "domain" DHCP option thereby potentially setting the
202 #    domain of all systems configured by DHCP
203 # 3) Provides the domain part for "expand-hosts"
204 #domain=thekelleys.org.uk
205 domain=somedomain.net
206
207 # Set Listen address
208 interface=${iflan} # listen on this interface (lo is already default)(no quotes to be used)
209 EOF
210
211 printf "\n* Restaring DNSMasq."
212 sudo systemctl restart dnsmasq
213
214 printf "\n* Installing IPSEC components."
215
216 if exists ipsec
217 then
218    printf "\n* Strongswan found, no installation needed."
219 else
220    sudo apt-get install strongswan -y
221 fi
222
223 printf "\n* Installing strongswan lib plugins."
224
225 if exists libstrongswan-extra-plugins
226 then
227    printf "\n* Strongswan plugins found, no installation needed."
228 else
229    sudo apt-get install libstrongswan-extra-plugins -y # may already be installed
230 fi
231
232 printf "\n* Installing charon lib plugins."
233
234 if exists libcharon-extra-plugins
235 then
236    printf "\n* Charon plugins found, no installation needed."
237 else
238    sudo apt-get install libcharon-extra-plugins -y  # may already be installed
239 fi
240
241 # add /etc/resolv.conf    rw, to usr.lib.ipsec.charon, under /etc/resolv.conf
242 if ! grep -q /etc/resolv.conf /etc/apparmor.d/usr.lib.ipsec.charon
243    then
244    # add /etc/resolv.conf    rw, to usr.lib.ipsec.charon, under /etc/resolv.conf
245    printf "\n* Resolve apparmor issue due to strongswan not allowed to write resolv.conf\n"
246    printf "\n* Adding resolv.conf edit right to charon\n"
247    sed -i -e '/\ \{2\}\/var\/lib\/strongswan\/\*\s\{5,\}/a \ \ \/etc\/resolv.conf\          rw,' /etc/apparmor    .d/usr.lib.ipsec.charon
248 fi
249
250
251 printf "\n* Install wget" # needed to download Proton certificate
252 if exists wget
253 then
254    printf "\n* wget is already installed."
255 else
256    sudo apt-get install wget -y
257 fi
258
259 printf "\n* Installing Proton certificate"
260 if test -f "/etc/ipsec.d/cacerts/protonvpn.der"
261 then
262    printf "\n* Certificate already installed, no action."
263 else
264    wget https://protonvpn.com/download/ProtonVPN_ike_root.der -O /tmp/protonvpn.der
265    sudo mv /tmp/protonvpn.der /etc/ipsec.d/cacerts/
266 fi
267
268 printf "\n* Writing the IPSEC configuration file."
269 cat << EOF > /etc/ipsec.conf
270 # ipsec.conf - strongSwan IPsec configuration file
271
272 # basic configuration
273
274 config setup
275    # strictcrlpolicy=yes
276    # uniqueids = no
277
278 conn ch03
279    left=%defaultroute
280    leftsourceip=%config
281    leftauth=eap-mschapv2
282    eap_identity=SomeUserName
283    right=185.2.3.4
284    rightsubnet=0.0.0.0/0
285    rightauth=pubkey
286    rightid=%185.2.3.4
287    rightca=/etc/ipsec.d/cacerts/protonvpn.der
288    rightca=/etc/ipsec.d/cacerts/protonvpn.der
289    keyexchange=ikev2
290    type=tunnel
291    auto=add
292 EOF
293
294 printf "\n* Writing the credentials file."
295 cat << EOF > /etc/ipsec.secrets
296 # This file holds shared secrets or RSA private keys for authentication.
297
298 # RSA private key for this host, authenticating it to any other host
299 # which knows the public part.
300
301 SomeUserName : EAP SomePassWord
302 EOF
303
308 printf "\n* Done.\n"

```

- The log for starting the connection. Last three lines are only found on the terminal if I stop the VPN connection shortly after starting.

udo ipsec up ch03
initiating IKE_SA ch03[1] to 185.2.3.4
generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
sending packet: from 192.2.3.4[500] to 185.2.3.4[500] (1128 bytes)
received packet: from 185.2.3.4[500] to 192.2.3.4[500] (38 bytes)
parsed IKE_SA_INIT response 0 [ N(INVAL_KE) ]
peer didn't accept DH group ECP_256, it requested ECP_384
initiating IKE_SA ch03[1] to 185.2.3.4
generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
sending packet: from 192.2.3.4[500] to 185.2.3.4[500] (1160 bytes)
received packet: from 185.2.3.4[500] to 192.2.3.4[500] (304 bytes)
parsed IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
selected proposal: IKE:AES_GCM_16_256/PRF_HMAC_SHA2_384/ECP_384
local host is behind NAT, sending keep alives
remote host is behind NAT
sending cert request for "C=CH, O=ProtonVPN AG, CN=ProtonVPN Root CA"
no IDi configured, fall back on IP address
establishing CHILD_SA ch03{1}
generating IKE_AUTH request 1 [ IDi N(INIT_CONTACT) CERTREQ CPRQ(ADDR DNS) SA TSi TSr N(MOBIKE_SUP) N(ADD_4_ADDR) N(MULT_AUTH) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
sending packet: from 192.2.3.4[4500] to 185.2.3.4[4500] (310 bytes)
received packet: from 185.2.3.4[4500] to 192.2.3.4[4500] (1248 bytes)
parsed IKE_AUTH response 1 [ EF(1/3) ]
received fragment #1 of 3, waiting for complete IKE message
received packet: from 185.2.3.4[4500] to 192.2.3.4[4500] (1165 bytes)
parsed IKE_AUTH response 1 [ EF(3/3) ]
received fragment #3 of 3, waiting for complete IKE message
received packet: from 185.2.3.4[4500] to 192.2.3.4[4500] (1248 bytes)
parsed IKE_AUTH response 1 [ EF(2/3) ]
received fragment #2 of 3, reassembled fragmented IKE message (3535 bytes)
parsed IKE_AUTH response 1 [ IDr CERT CERT AUTH EAP/REQ/ID ]
received end entity cert "CN=node-ch-02.protonvpn.net"
received issuer cert "C=CH, O=ProtonVPN AG, CN=ProtonVPN Intermediate CA 1"
  using certificate "CN=node-ch-02.protonvpn.net"
  using untrusted intermediate certificate "C=CH, O=ProtonVPN AG, CN=ProtonVPN Intermediate CA 1"
checking certificate status of "CN=node-ch-02.protonvpn.net"
certificate status is not available
  using trusted ca certificate "C=CH, O=ProtonVPN AG, CN=ProtonVPN Root CA"
checking certificate status of "C=CH, O=ProtonVPN AG, CN=ProtonVPN Intermediate CA 1"
certificate status is not available
  reached self-signed root ca with a path length of 1
authentication of '185.2.3.4' with RSA_EMSA_PKCS1_SHA2_384 successful
server requested EAP_IDENTITY (id 0x00), sending 'SomeUserName'
generating IKE_AUTH request 2 [ EAP/RES/ID ]
sending packet: from 192.2.3.4[4500] to 185.2.3.4[4500] (93 bytes)
received packet: from 185.2.3.4[4500] to 192.2.3.4[4500] (83 bytes)
parsed IKE_AUTH response 2 [ EAP/REQ/MD5 ]
server requested EAP_MD5 authentication (id 0x01)
requesting EAP_MSCHAPV2 authentication, sending EAP_NAK
generating IKE_AUTH request 3 [ EAP/RES/NAK ]
sending packet: from 192.2.3.4[4500] to 185.2.3.4[4500] (67 bytes)
received packet: from 185.2.3.4[4500] to 192.2.3.4[4500] (104 bytes)
parsed IKE_AUTH response 3 [ EAP/REQ/MSCHAPV2 ]
server requested EAP_MSCHAPV2 authentication (id 0x02)
generating IKE_AUTH request 4 [ EAP/RES/MSCHAPV2 ]
sending packet: from 192.2.3.4[4500] to 185.2.3.4[4500] (147 bytes)
received packet: from 185.2.3.4[4500] to 192.2.3.4[4500] (112 bytes)
parsed IKE_AUTH response 4 [ EAP/REQ/MSCHAPV2 ]
EAP-MS-CHAPv2 succeeded: '(null)'
generating IKE_AUTH request 5 [ EAP/RES/MSCHAPV2 ]
sending packet: from 192.2.3.4[4500] to 185.2.3.4[4500] (67 bytes)
received packet: from 185.2.3.4[4500] to 192.2.3.4[4500] (65 bytes)
parsed IKE_AUTH response 5 [ EAP/SUCC ]
EAP method EAP_MSCHAPV2 succeeded, MSK established
authentication of '192.2.3.4' (myself) with EAP
generating IKE_AUTH request 6 [ AUTH ]
sending packet: from 192.2.3.4[4500] to 185.2.3.4[4500] (113 bytes)
received packet: from 185.2.3.4[4500] to 192.2.3.4[4500] (341 bytes)
parsed IKE_AUTH response 6 [ AUTH CPRP(ADDR DNS MASK) SA TSi TSr N(MOBIKE_SUP) N(ADD_4_ADDR) N(ADD_4_ADDR) N(ADD_4_ADDR) N(ADD_4_ADDR) N(ADD_4_ADDR) N(ADD_4_ADDR) N(ADD_4_ADDR) N(ADD_4_ADDR) ]
authentication of '185.2.3.4' with EAP successful
IKE_SA ch03[1] established between 192.2.3.4[192.2.3.4]...185.2.3.4[185.2.3.4]
scheduling reauthentication in 10154s
maximum IKE_SA lifetime 10694s
installing DNS server 10.1.0.111 to /etc/resolv.conf
handling INTERNAL_IP4_NETMASK attribute failed
installing new virtual IP 10.1.1.123
selected proposal: ESP:AES_CBC_256/HMAC_SHA2_256_128/NO_EXT_SEQ
CHILD_SA ch03{1} established with SPIs ceddbe61_i ce998c25_o and TS 10.1.1.123/32 === 0.0.0.0/0
connection 'ch03' established successfully

**Comment by tobiasbrunner** [comment]
> My SSH connection to the server (or vpn client if you may) is interrupted as soon as the connection establishes, and this prevents me from completing the setup.

The VPN tunnel covers all traffic (`0.0.0.0/0`). So configure [passthrough policies](https://wiki.strongswan.org/projects/strongswan/wiki/UsableExamples#Passthrough-policy) for SSH.

> In the connection log I find "handling INTERNAL_IP4_NETMASK attribute failed" after setting the dns server in /etc/resolv.conf which I am not sure how to solv.

That attribute is not used by strongSwan, shouldn't be a problem.

> The connection seems to work but forwarding the local network fails, although due to nr 1) I have some difficulty to find out why.

What do you mean with "forwarding the local network" mean? What exactly is your goal? Note that your VPN server assigns a dynamic virtual IP address. So if your goal is to send traffic from hosts behind this hosts via VPN, you have to create a custom updown script that installs NAT rules that maps such traffic to the virtual IP address so it gets tunneled ([this test scenario](https://www.strongswan.org/testing/testresults/ikev2/nat-virtual-ip/) illustrates this, uses the script [here](https://github.com/strongswan/strongswan/blob/master/testing/tests/ikev2/nat-virtual-ip/hosts/moon/etc/nat_updown)).

**Comment by hombibi** [comment]
Thanks Tobias, that helps me another step forward: I got the SSH connection to persist, but there are issues.
The ipsec connection is now:
```
257 conn ch03
258    left=%defaultroute
259    leftsourceip=%config
260    leftauth=eap-mschapv2
261    eap_identity=SomeUserName
262    right=185.2.3.4
263    rightsubnet=0.0.0.0/0
264    rightauth=pubkey
265    rightid=%185.2.3.4
266    rightca=/etc/ipsec.d/cacerts/protonvpn.der
267    keyexchange=ikev2
268    type=tunnel
269    auto=add
270 conn pass
271    # makes sure those conns are excluded from every conn selection
272    left=127.0.0.1
273    #leftsubnet=%dynamic[tcp/22]
274    leftsubnet=192.6.7.0/24
275    #rightsubnet=0.0.0.0/0
276    rightsubnet=192.6.7.0/24
277    type=passthrough
278    auto=route
```
I further took out the nameserver 1.1.1.1 from the dnsmasq config section, to ensure only the one provided by the vpn connection in /etc/resolv.conf

After various tests, the vpn connection itself seems fine: Executed from the ubuntu server traceroute shows traffic is directed through the tunnel, and a dns leak (script) test shows only the expected dns servers as provided by the tunnel.

However with the passthrough settings as indicated a device set with GW and DNS to the ubuntu vpn client shows that dns leaks and that internet access bypasses the vpn tunnel. Taking out the passthrough settings however, the device has no access to the vpn tunnel and subsquent network.

When I stop the VPN connection, I can again ping GW and DNS, and adding a nameserver to /etc/resolv. conf makes the router work once more.

It seems the passthrough setup is not correct, I searched for examples setups, but could not find much information how to interpret what values I should use. The provided example does not resolve the SSH interruption.

Also the router configuration once the vpn is up is missing something, since packages from lan devices seen not be forewarded, or returned. For this I still need to have a look at the scripts you referenced. thanks for that.

**Reply by tobiasbrunner** [reply]
As mentioned, you'll need a NAT rule (dynamically installed via updown script as it depends on the virtual IP assigned by the VPN server) in order to tunnel the forwarded traffic. Usually, strongSwan enforces the virtual IP via a special route in table 220. But with the NAT rule that's not required and could potentially cause conflicts if the remote subnet is `0.0.0.0/0`, so disabling `charon.install_routes` in strongswan.conf might be a good idea.
